ocaml/xapi/xapi_cluster_host.ml

```diff
@@ -65,76 +46,80 @@ let call_api_function_with_alert ~__context ~msg ~cls ~obj_uuid ~body
+      assert_pif_attached_to ~host ~pIF ~__context;
       let ref = Ref.make () in
-      let dbg = Context.string_of_task __context in
       let uuid = Uuidm.to_string (Uuidm.create `V4) in
+      let pifrec = Db.PIF.get_record ~__context ~self:pIF in
+      assert_pif_prerequisites (pIF,pifrec);
+      Db.Cluster_host.create ~__context ~ref ~uuid ~cluster ~host ~pIF ~enabled:false
+        ~current_operations:[] ~allowed_operations:[] ~other_config:[] ~joined:false;
+
+      ref
+  )
+
+(* Helper function atomically enables clusterd and joins the cluster_host *)
+let join_internal ~__context ~self =
+  with_clustering_lock (fun () ->
+      let dbg = Context.string_of_task __context in
+      let cluster = Db.Cluster_host.get_cluster ~__context ~self in
       let cluster_token = Db.Cluster.get_cluster_token ~__context ~self:cluster in
-      let pifref,pifrec = pif,(Db.PIF.get_record ~__context ~self:pif) in
-      assert_pif_prerequisites (pifref,pifrec);
+      let pifref = Db.Cluster_host.get_PIF ~__context ~self in
+      let pifrec = Db.PIF.get_record ~__context ~self:pifref in
       let ip = ip_of_pif (pifref,pifrec) in
-      let ip_list = List.map (fun cluster_host ->
-        let pref = Db.Cluster_host.get_PIF ~__context ~self:cluster_host in
+      let ip_list = List.map (fun self ->
+        let pref = Db.Cluster_host.get_PIF ~__context ~self in
         let prec = Db.PIF.get_record ~__context ~self:pref in
           ip_of_pif (pref,prec)
         ) (Db.Cluster.get_cluster_hosts ~__context ~self:cluster) in
+
+      debug "Enabling clusterd and joining cluster_host %s" (Ref.string_of self);
       Xapi_clustering.Daemon.enable ~__context;
-      let result = Cluster_client.LocalClient.join (rpc ~__context) dbg cluster_token ip ip_list in
-      match result with
+      Db.Cluster_host.set_joined ~__context ~self ~value:true;
+      Db.Cluster_host.set_enabled ~__context ~self ~value:true;
+      Cluster_client.LocalClient.join (rpc ~__context) dbg cluster_token ip ip_list
+  )
+
+(* Enable cluster_host in client layer via clusterd *)
+let resync_host ~__context ~host =
+  match (find_cluster_host ~__context ~host) with
+  | None      -> () (* no clusters exist *)
+  | Some self ->    (* cluster_host and cluster exist *)
+    if not (Db.Cluster_host.get_joined ~__context ~self) then begin
+       match join_internal ~__context ~self with
       | Result.Ok () ->
-        Db.Cluster_host.create ~__context ~ref ~uuid ~cluster ~host ~pIF:pifref ~enabled:true
-          ~current_operations:[] ~allowed_operations:[] ~other_config:[];
-        debug "Cluster_host.create was successful; cluster_host: %s" (Ref.string_of ref);
-        ref
+        debug "Cluster_host.create was successful; cluster_host: %s" (Ref.string_of self)
       | Result.Error error ->
         warn "Error occurred during Cluster_host.create";
         handle_error error
-    )
+    end;
+
+    if Db.Cluster_host.get_enabled ~__context ~self then begin
```

Comment:
Sorry, one more change here: if we have just joined we should not try to go and enable it immediately: it will fail with other operation in progress (we are the other operation: Cluster.add).

So there should be an `else` here.